File: 3/3.hs
import qualified Data.Set as Set

main :: IO ()
main = print $ maximum $ primeFactors 600851475143

primeFactors :: Int -> [Int]
primeFactors n =
    Set.toList $
    foldr (\f set -> if n `mod` f == 0
                       then let subFactors = primeFactors f
                            in if Set.null $ Set.fromList subFactors
                                 then Set.insert f set
                                 else set `Set.union` Set.fromList subFactors
                       else set)
          Set.empty
          [2 .. (floor . (sqrt::Double->Double) . fromIntegral) n]

File: 3/3.rs
use std::f64::floor;
use std::hashmap::HashSet;
use std::iter::range_inclusive;
use std::num::sqrt;

fn main() {
    let factors = prime_factors(600851475143 as f64);
    println!("{:f}", *factors.iter().max().unwrap());
}

fn prime_factors(n: f64) -> ~[f64] {
    let mut factors: HashSet<f64> = HashSet::new();
    for i in range_inclusive(2 as f64, sqrt(n)) {
        if n%floor(i) == 0 as f64 {
            let sub_factors = prime_factors(i);
            for &subf in sub_factors.iter() { factors.insert(subf); }
        }
    }
    if factors.is_empty() {
        return ~[n];
    }
    factors.iter().map(|&f| f).collect()
}
